File: 3213-count-subarrays-where-max-element-appears-at-least-k-times/count-subarrays-where-max-element-appears-at-least-k-times.java
class Solution {
    public long countSubarrays(int[] nums, int k) {
        int n = nums.length;
        int maxVal = 0;
        for (int num : nums) {
            maxVal = Math.max(maxVal, num);
        }

        long result = 0;
        int count = 0;
        int left = 0;

        for (int right = 0; right < n; right++) {
            if (nums[right] == maxVal) {
                count++;
            }

            while (count >= k) {
                if (nums[left] == maxVal) {
                    count--;
                }
                
                left++;
            }

            result += left;
        }

        return result;
    }
}

File: 3213-count-subarrays-where-max-element-appears-at-least-k-times/count-subarrays-where-max-element-appears-at-least-k-times.go
func countSubarrays(nums []int, k int) int64 {
    n := len(nums);
    maxVal := 0;
    for _, num := range nums {
        maxVal = max(maxVal, num);
    }

    var result int64;
    count, left := 0, 0;

    for right := 0; right < n ; right++ {
        if nums[right] == maxVal {
            count++;
        }

        for count >= k {
            if nums[left] == maxVal {
                count--;
            }

            left++;
        }

        result += int64(left);
    }

    return result;
}

func max(a, b int) int {
    if a > b {
        return a;
    }

    return b;
}